### PR TITLE
fix #250: introduce tmpdir option for overriding the system tmp dir

### LINKED ISSUE #250: should not lock dir in os.tmp
## Operating System
- [ ] Linux
- [ ] Windows 7
- [ ] Windows 10
- [ ] MacOS
- other: 

## NodeJS Version
- [ ] 0.x
- [ ] 4.x
- [ ] 6.x
- [ ] 7.x
- other: 

## Tmp Version
0.2.0

## Expected Behavior

should allow any dir

## Experienced Behavior
TBD:What did actually happen?

> Error: dir option must be relative to "C:\Users\User\AppData\Local\Temp", found "T:\cache\yarn-cache\tmp".

### PR DESCRIPTION
**Pull request by silkentrance**
fix #250: introduce tmpdir option for overriding the system tmp dir

also fix issue with changes to the options leaking out, now copying the options to a new object

### CODE REVIEW

File: docs/tmp.js.html [hunk 1]
```diff
@@ -132,17 +132,18 @@ <h1 class="page-title">Source: tmp.js</h1>
...
+    name = _generateTmpName(opts);
     try {
       fs.statSync(name);
     } catch (e) {
       return name;
     }
   } while (tries-- > 0);
 
-  throw new Error('Could not get a unique tmp filename, max tries reached');
+  throw new Error('Could not get a unique tmp filename, max tries reached ' + name);
```
Comment (raszi) [inline docs/tmp.js.html:146]: Why this change was necessary? I believe it is not too useful to return a random name in the error message.

File: docs/tmp.js.html [hunk 2]
```diff
@@ -491,7 +498,7 @@ <h1 class="page-title">Source: tmp.js</h1>
  */
 function _generateTmpName(opts) {
 
-  const tmpDir = _getTmpDir();
+  const tmpDir = opts.tmpdir;
```
Comment (raszi) [inline docs/tmp.js.html:501]: I would simply pass the options to the `_getTmpDir` function and return the required directory from there.

File: docs/tmp.js.html [hunk 3]
```diff
@@ -523,10 +529,14 @@ <h1 class="page-title">Source: tmp.js</h1>
...
-  const tmpDir = _getTmpDir();
+
+  options.tmpdir = _isUndefined(options.tmpdir) ? _getTmpDir() : _resolvePath(options.tmpdir, options.tmpdir);
+
+  const tmpDir = options.tmpdir;
+
   /* istanbul ignore else */
   if (!_isUndefined(options.name))
-    _assertIsRelative(options.name, 'name', tmpDir);
+    _assertIsRelative(options.name, 'name', tmpDir)
```
Comment (raszi) [inline docs/tmp.js.html:539]: Missing semicolon.

File: docs/tmp.js.html (hunk 1 above)
Comment (silkentrance) [reply docs/tmp.js.html:146]: ups, temporary change for debugging purposes... will undo

File: docs/tmp.js.html (hunk 2 above)
Comment (silkentrance) [reply docs/tmp.js.html:501]: done

File: docs/tmp.js.html (hunk 3 above)
Comment (silkentrance) [reply docs/tmp.js.html:539]: done

**Comment by silkentrance** [comment]
A new release needs to be made available with an updated version of the changelog.

See the instructions in #253 on how to do that.

**Comment by silkentrance** [comment]
@raszi I have already bumped the version and updated the changelog. Have fun releasing 0.2.1 :)

**Pull request merged**